### PR TITLE
New random_sampler aggregation for sampling documents in aggregations

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new sampling aggregation that performs a background sampling over all documents in an index. 

The syntax is as follows:
```
{
  "aggregations": {
    "sampling": {
      "random_sampler": {
        "probability": 0.1
      },
      "aggs": {
        "price_percentiles": {
          "percentiles": {
            "field": "taxful_total_price"
          }
        }
      }
    }
  }
}
```

This aggregation provides fast random sampling over the entire document set in order to speed up costly aggregations.

Testing this over a variety of aggregations and data sets, the median speed up when sampling at `0.001` over millions of documents is around 70X speed improvement.

image

Relative error rate does rely on the size of the data and the aggregation kind. Here are some typically expected numbers when sampling over 10s of millions of documents. `p` is the configured probability and `n` is the number of documents matched by your provided filter query.

image

### CODE REVIEW

File: docs/reference/aggregations/bucket/categorize-text-aggregation.asciidoc [hunk 1]
```diff
@@ -13,8 +13,9 @@ tokens are used to categorize the text.
 
 NOTE: If you have considerable memory allocated to your JVM but are receiving circuit breaker exceptions from this
       aggregation, you may be attempting to categorize text that is poorly formatted for categorization. Consider
-      adding `categorization_filters` or running under <<search-aggregations-bucket-sampler-aggregation,sampler>> or
-      <<search-aggregations-bucket-diversified-sampler-aggregation,diversified sampler>> to explore the created categories.
+      adding `categorization_filters` or running under <<search-aggregations-bucket-sampler-aggregation,sampler>>,
+      <<search-aggregations-bucket-diversified-sampler-aggregation,diversified sampler>>, or
+      <<search-random-sampler-aggregation, random sampler>> to explore the created categories.
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/categorize-text-aggregation.asciidoc:18]: This link is invalid and causes the following doc build error: 
> 10:38:57 INFO:build_docs:asciidoctor: WARNING: invalid reference: search-random-sampler-aggregation
I suspect you want to link to this page instead:
```suggestion
      <<search-aggregations-random-sampler-aggregation,random sampler>> to explore the created categories.
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+provided, a document is returned if it is matched by the query and if the
+document is in the random sampling. The sampling is not done over the matched
+documents.
+
+Consider the set of documents `[1, 2, 3, 4, 5]`. Your query matches `[1, 3, 5]`
+and the randomly sampled set is `[2, 4, 5]`. In this case, the document returned
+would be `[5]`.
+
+This type of sampling provides almost linear performance in relation to sampling
+probability.
```
Comment (tveasey) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:72]: I know you qualify this below, but I think I would reword this slightly so it is a bit more explicit about the performance improvement. Also strictly linear in probability would imply proportional to p not 1 / p, so I would instead reference the reduction in sample document size.
```suggestion
This type of sampling provides almost linear improvement in query latency in relation to the amount
by which sampling reduces the document set size.
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+image::images/aggregations/random-sampler-agg-graph.png[Graph of the median speedup by sampling factor,align="center"]
+
+Above is the typically seen speed up for the majority of aggregations. For certain aggregations, the speed up may not
```
Comment (tveasey) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:76]:
```suggestion
Above is typical of the speed up for the majority of aggregations for a test data set of 63 million documents. The exact constants will depend on the data set size and the number of shards, but the form of the relationship between speed up and probability holds widely. For certain aggregations, the speed up may not
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+be as dramatic. These aggregations have some constant overhead unrelated to the number of documents seen. Even for
+those aggregations, the speed improvements can be significant.
+
+This sample is generated by sampling values from the geometric distribution
+(`(1-p)^(k-1)*p`) at the provided `probability` (`p` in distribution equation).
+The values returned from the distribution indicate how many documents to skip in
+the background. It follows that the expected success from the distribution is
```
Comment (tveasey) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:83]:
```suggestion
the background. This is equivalent to selecting documents uniformly at random. It follows that the expected number of failures before a success is
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+Consequently, if you had only 80 documents in your index or matched by your
+filter, you would most likely receive no results.
+
+image::images/aggregations/relative-error-vs-doc-count.png[Graph of the relative error by sampling probability and doc count,align="center"]
+
+In the above image `p` is the probability provided to the aggregation, and `n` is the number of documents matched by whatever
+query is provided. One can see the impact of outliers on `sum` and `mean`, but when many documents are still matched at
+higher sampling rates, the relative error is still low.
+
+NOTE: This data is empirical and represents aggregations against a real-world dataset. The variance in your own data may
```
Comment (tveasey) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:95]:
```suggestion
NOTE: This represents the result of aggregations against a typical positively skewed APM data set which also has outliers in the upper tail. The linear dependence of the relative error on the sample size is found to hold widely, but the slope depends on the variation in the quantity being aggregated. As such, the variance in your own data may
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+probability.
+
+image::images/aggregations/random-sampler-agg-graph.png[Graph of the median speedup by sampling factor,align="center"]
+
+Above is the typically seen speed up for the majority of aggregations. For certain aggregations, the speed up may not
+be as dramatic. These aggregations have some constant overhead unrelated to the number of documents seen. Even for
+those aggregations, the speed improvements can be significant.
+
+This sample is generated by sampling values from the geometric distribution
+(`(1-p)^(k-1)*p`) at the provided `probability` (`p` in distribution equation).
```
Comment (tveasey) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:81]:
```suggestion
This sample set is generated by skipping documents using a geometric distribution
(`(1-p)^(k-1)*p`) with success probability being the provided `probability` (`p` in distribution equation).
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+The values returned from the distribution indicate how many documents to skip in
+the background. It follows that the expected success from the distribution is
+`(1-p)/p`. For example, with the `"probability": 0.01`, the expected success (or
```
Comment (tveasey) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:84]:
```suggestion
`(1-p)/p`. For example, with the `"probability": 0.01`, the expected number of failures (or
```

**Review by tveasey — COMMENTED** [review]
Limits and new tests look good. I made some suggestions on some of the wording in the documentation.

**Review by tveasey — APPROVED** [review]

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+Consider the set of documents `[1, 2, 3, 4, 5]`. Your query matches `[1, 3, 5]`
+and the randomly sampled set is `[2, 4, 5]`. In this case, the document returned
+would be `[5]`.
+
+This type of sampling provides almost linear improvement in query latency in relation to the amount
+by which sampling reduces the document set size.
+
+image::images/aggregations/random-sampler-agg-graph.png[Graph of the median speedup by sampling factor,align="center"]
+
+Above is typical of the speed up for the majority of aggregations for a test data set of 63 million documents. The exact constants will depend on the data set size and the number of shards, but the form of the relationship between speed up and probability holds widely. For certain aggregations, the speed up may not
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:76]:
```suggestion
This graph is typical of the speed up for the majority of aggregations for a test data set of 63 million documents. The exact constants will depend on the data set size and the number of shards, but the form of the relationship between speed up and probability holds widely. For certain aggregations, the speed up may not
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+be as dramatic. These aggregations have some constant overhead unrelated to the number of documents seen. Even for
+those aggregations, the speed improvements can be significant.
+
+This sample set is generated by skipping documents using a geometric distribution
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:80]:
```suggestion
The sample set is generated by skipping documents using a geometric distribution
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+provided, a document is returned if it is matched by the query and if the
+document is in the random sampling. The sampling is not done over the matched
+documents.
+
+Consider the set of documents `[1, 2, 3, 4, 5]`. Your query matches `[1, 3, 5]`
+and the randomly sampled set is `[2, 4, 5]`. In this case, the document returned
+would be `[5]`.
+
+This type of sampling provides almost linear improvement in query latency in relation to the amount
+by which sampling reduces the document set size.
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:72]: It seems like this statement summarizes the information in the graph; if so, let's add a colon to denote that relationship:
```suggestion
by which sampling reduces the document set size:
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+image::images/aggregations/random-sampler-agg-graph.png[Graph of the median speedup by sampling factor,align="center"]
+
+Above is typical of the speed up for the majority of aggregations for a test data set of 63 million documents. The exact constants will depend on the data set size and the number of shards, but the form of the relationship between speed up and probability holds widely. For certain aggregations, the speed up may not
+be as dramatic. These aggregations have some constant overhead unrelated to the number of documents seen. Even for
+those aggregations, the speed improvements can be significant.
+
+This sample set is generated by skipping documents using a geometric distribution
+(`(1-p)^(k-1)*p`) with success probability being the provided `probability` (`p` in distribution equation).
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:81]:
```suggestion
(`(1-p)^(k-1)*p`) with success probability being the provided `probability` (`p` in the distribution equation).
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+The values returned from the distribution indicate how many documents to skip in
+the background. This is equivalent to selecting documents uniformly at random. It follows that the expected number of failures before a success is
+`(1-p)/p`. For example, with the `"probability": 0.01`, the expected number of failures (or
+average number of documents skipped) would be `99` with a variance of `9900`.
+Consequently, if you had only 80 documents in your index or matched by your
+filter, you would most likely receive no results.
+
+image::images/aggregations/relative-error-vs-doc-count.png[Graph of the relative error by sampling probability and doc count,align="center"]
+
+In the above image `p` is the probability provided to the aggregation, and `n` is the number of documents matched by whatever
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:91]: Not mandatory, but I think it would be nice to move this paragraph before the image, so that it provides context  in advance.

If you move this paragraph, it needs minor changes like this:
```suggestion
In the following graph, `p` is the probability provided to the aggregation, and `n` is the number of documents matched by whatever
```

File: docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+query is provided. One can see the impact of outliers on `sum` and `mean`, but when many documents are still matched at
```
Comment (lcawl) [inline docs/reference/aggregations/bucket/random-sampler-aggregation.asciidoc:92]:
```suggestion
query is provided. You can see the impact of outliers on `sum` and `mean`, but when many documents are still matched at
```

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java [hunk 14]
```diff
@@ -163,7 +169,7 @@ public String getType() {
 
     @Override
     public Version getMinimalSupportedVersion() {
-        return Version.V_8_1_0;
+        return Version.V_8_2_0;
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java:172]: What's the policy around setting this WRT experimental features?  I'm not totally sure how this is used.

**Review by not-napoleon — APPROVED** [review]
LGTM.  Excited to let folks start trying this out.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java (hunk 14 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java:172]: @not-napoleon I think the main use is failing early on CCS? Regardless, its a required override and to serialize, we need at least 8.2.0 :).

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by lcawl — APPROVED** [review]
Docs LGTM
